src/state/slices/opportunitiesSlice/resolvers/osmosis/index.ts

```diff
@@ -0,0 +1,114 @@
+import type { ToAssetIdArgs } from '@shapeshiftoss/caip'
+import { osmosisChainId, toAssetId } from '@shapeshiftoss/caip'
+import { DefiProvider, DefiType } from 'features/defi/contexts/DefiManagerProvider/DefiCommon'
+import type { ReduxState } from 'state/reducer'
+import type { PortfolioAccountBalancesById } from 'state/slices/portfolioSlice/portfolioSliceCommon'
+import {
+  selectAssetById,
+  selectPortfolioAccountBalances,
+  selectPortfolioLoadingStatusGranular,
+} from 'state/slices/selectors'
+
+import type {
+  GetOpportunityIdsOutput,
+  GetOpportunityMetadataOutput,
+  LpId,
+  OpportunityMetadata,
+} from '../../types'
+import { toOpportunityId } from '../../utils'
+import type { OpportunitiesMetadataResolverInput, OpportunityUserDataResolverInput } from '../types'
+import { generateAssetIdFromOsmosisDenom, getPools } from './utils'
+
+export const osmosisLpOpportunitiesMetadataResolver = async ({
+  opportunityType,
+  reduxApi,
+}: OpportunitiesMetadataResolverInput): Promise<{ data: GetOpportunityMetadataOutput }> => {
+  const { getState } = reduxApi
+  const state: any = getState()
+
+  const lpOpportunitiesById: Record<LpId, OpportunityMetadata> = {}
+
+  const liquidityPools = await getPools()
+
+  for (const pool of liquidityPools) {
+    const toAssetIdParts: ToAssetIdArgs = {
+      assetNamespace: 'ibc',
+      assetReference: `gamm/pool/${pool.id}`,
+      chainId: osmosisChainId,
+    }
+
+    const assetId = toAssetId(toAssetIdParts)
+    const opportunityId = toOpportunityId(toAssetIdParts)
+    const asset = selectAssetById(state, assetId)
+
+    if (!asset) continue
+
+    lpOpportunitiesById[opportunityId] = {
+      apy: pool.apy,
+      assetId,
+      provider: DefiProvider.Osmosis,
+      tvl: pool.tvl,
+      type: DefiType.LiquidityPool,
+      underlyingAssetId: assetId,
+      underlyingAssetIds: [
+        generateAssetIdFromOsmosisDenom(pool.pool_assets[0].token.denom),
+        generateAssetIdFromOsmosisDenom(pool.pool_assets[1].token.denom),
+      ],
+      underlyingAssetRatiosBaseUnit: [
+        pool.pool_assets[0].token.amount,
+        pool.pool_assets[1].token.amount,
+      ],
+      name: pool.name,
+    }
+  }
+
+  const data = {
+    byId: lpOpportunitiesById,
+    type: opportunityType,
+  }
+
+  return { data }
+}
+
+export const osmosisLpUserDataResolver = ({
```

Comment:
Confirmed these exist in portfolio and having userDataResolver only as a guard is the correct approach!